contrib/tinks/src/game/world.nit

```diff
@@ -0,0 +1,276 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Physical world logic
+module world is serialize
+
+import noise
+import more_collections
+
+import framework
+
+redef class TGame
+	# The physical world in which this game happens
+	var world = new World(self)
+end
+
+# A terrain features (a rock, a tree, etc.)
+class Feature
+	super Ruled
+	redef type R: FeatureRule
+
+	# Top-left corner of this feature
+	var pos: Pos
+end
+
+# Metadata for a `Feature`
+class FeatureRule
+	super Rule
+
+	# Strength to resist on `World::explode`
+	var strength: Int
+end
+
+# The physical world of the `game`
+class World
+	# Associated `TGame`
+	var game: TGame
+
+	# Past blast sites
+	var blast_sites = new Array[Pos]
+
+	private var mountain_map: Noise do
+		var map = new PerlinNoise
+		map.period = 10.0
+		return map
+	end
+
+	private var forest_map: Noise do
+		var map = new PerlinNoise
+		map.period = 24.0
+		return map
+	end
+
+	# Cache of discovered features, also keeps tracks of changes
+	private var features_cache = new FeatureMap
+
+	# Get the `Feature` at `x, y`, if any
+	fun [](x, y: Int): nullable Feature
+	do
+		if features_cache.has(x, y) then return features_cache[x, y]
+
+		# Generate a feature from the noise map
+		var pos = new Pos(x.to_f+0.5, y.to_f+0.5)
+
+		var feature = null
+		if mountain_map[x.to_f, y.to_f] > 0.55 then
+			feature = new Feature(game.story.rock, pos)
+		else if forest_map[x.to_f, y.to_f] > 0.5 then
+			feature = new Feature(game.story.tree, pos)
+		end
+
+		# Update cache
+		features_cache[x, y] = feature
+
+		return feature
+	end
+
+	# Detect the first collision with a `Feature` from `src` to `dst`
+	#
+	# This is the main collision detection method used by tanks shots, and at tank movement.
+	# The idea is to check all cases between `src` and `dst` and return the first feature found.
+	# Returns `null` if there is no obstacle features.
+	#
+	# Example of the cases that would be checked between `s` and `d`:
+	#
+	# ~~~raw
+	# ................
+	# .s###...........
+	# ....######......
+	# .........####d..
+	# ................
+	# ~~~
```

Comment:
That link aims for a pretty output, that's not what I'm doing here. It's important for collision detection that the line doesn't skip tiles.